[SSL certs] Document v5 Windows proxy cases (#7641)
DataDog/documentation: content/en/agent/faq/certificate_verify_failed-error.md

1. Download the [Python script][1].
2. Run it in your local terminal or shell. The script works with Python 2 and 3.
Datadog US: `python find_agents_with_connectivity_problems.py --api-key API_KEY --application-key APPLICATION_KEY --site us`
Datadog EU: `python find_agents_with_connectivity_problems.py --api-key API_KEY --application-key APPLICATION_KEY --site eu`
3. Find the CSV output in `hosts_agents.csv`.

> Get the API and application key [from your account settings page][2]. If you are using the EU site, see your [EU account settings page][3].

Reviewer: This looks correct to me here, but the rendered page has the US url on the `EU account settings page` link

Reviewer 2: The region selector in the upper right controls US/EU site links.

Author: ok, I'll simply remove the mention of the EU site and link

Reviewer: @jeremy-lq in that case, we should remove `If you are using the EU site, see your [EU account settings page][3].`?

1. Download the [Python script][1].
2. Run it in your local terminal or shell. The script works with Python 2 and 3.
Datadog US: `python find_agents_with_connectivity_problems.py --api-key API_KEY --application-key APPLICATION_KEY --site us`
Datadog EU: `python find_agents_with_connectivity_problems.py --api-key API_KEY --application-key APPLICATION_KEY --site eu`
3. Find the CSV output in `hosts_agents.csv`.

Get the API and application key [from your account settings page][2].

### Fixing by upgrading to Agent 5.32.7

If you are currently running Agent 5.x on a 64-bit host, Datadog recommends upgrading to Agent 5.32.7+. This ensures that the Agent continues to function in a variety of different scenarios, with the minimum amount of changes.

Centos/Red Hat: `sudo yum check-update && sudo yum install datadog-agent`
Debian/Ubuntu: `sudo apt-get update && sudo apt-get install datadog-agent`
Windows (from versions > 5.12.0): Download the Datadog [Agent installer][3]. `start /wait msiexec /qn /i ddagent-cli-latest.msi`

> More platforms and configuration management options detailed [here][4].

Reviewer 3: ```suggestion
More platforms and configuration management options detailed [on the Agent install page][4].
```

restart-service -Force datadogagent
```

If your Agent is configured to use a proxy, see the additional step in the [dedicated section below](#additional-step-if-the-windows-agent-5-x-is-configured-to-use-a-proxy-or-the-curl-http-client).

*Using the Windows GUI*

Delete `datadog-cert.pem`. You can locate this file in:

* Agent `>=5.12.0`:
  * 64-bit Windows: `C:\Program Files\Datadog\Datadog Agent\agent\`
  * 32-bit Windows: The Datadog Agent was not available for 32-bit Windows systems starting in Agent 5.12
* Agent `<= 5.11.x`:

>   * 64-bit Windows: `C:\Program Files (x86)\Datadog\Datadog Agent\files\`

Reviewer 4: Is `C:\Program Files (x86)\Datadog\Datadog Agent\files\` correct for 64-bit Windows? 
Usually, `C:\Program Files (x86)\Datadog\Datadog Agent\files\` contains 32-bit applications.

  * 32-bit Windows: `C:\Program Files\Datadog\Datadog Agent\files\`

Once the file is removed, restart the Datadog Service from the Windows Service Manager.

> If your Agent is configured to use a proxy, see the additional step in the [dedicated section below](#additional-step-if-the-windows-agent-5-x-is-configured-to-use-a-proxy-or-the-curl-http-client).

Reviewer: It's not really an additional step, right? It's a different step? For windows agent 5.12+, if they're replacing datadog-cert.pem, they don't need to delete it first (but it does not matter if they do).

For Windows Agent v5, `<= 5.11.x`, do they need to delete the datadog-cert.pem? or is adding the ca_certs option enough

Reviewer 5: For Windows Agent `<= 5.11.x`, adding the `ca_certs` option is enough. We're only using `datadog-cert.pem` if this option is not set, as seen in this code block:
https://github.com/DataDog/dd-agent/blob/4d2b6eb26fc2583f24fd992463452ef555702231/config.py#L667-L670

(This code block hasn't changed since it was introduced in `3.6.0`)

Author: yes, that's fair, I'll change this

Note: `datadog.conf` is located in `C:\ProgramData\Datadog\datadog.conf`.

In this case, removing `datadog-cert.pem` will not allow the Agent to regain connectivity to Datadog. Instead, take this action:

* Windows Agent v5, `>= 5.12.0`: replace the `datadog-cert.pem` file with the version that is shipped in 5.32.7. Using the Powershell CLI:

```shell
Invoke-WebRequest -Uri f"https://raw.githubusercontent.com/DataDog/dd-agent/5.32.7/datadog-cert.pem" -OutFile "C:\Program Files\Datadog\Datadog Agent\agent\datadog-cert.pem"
restart-service -Force datadogagent
```

* Windows Agent v5, `<= 5.11.x`: set the following option in `datadog.conf` using the `Datadog Agent Manager` program provided by the Agent or by directly editing the `datadog.conf` file:
  * 64-bit Windows: `ca_certs: C:\Program Files (x86)\Datadog\Datadog Agent\files\ca-certificates.crt`
  * 32-bit Windows: `ca_certs: C:\Program Files\Datadog\Datadog Agent\files\ca-certificates.crt`
  After `datadog.conf` has been updated, restart the Datadog Service from the Windows Service Manager.

[1]: https://static.datadoghq.com/find_agents_with_connectivity_problems.py
[2]: https://app.datadoghq.com/account/settings#api

> [3]: https://s3.amazonaws.com/ddagent-windows-stable/ddagent-cli-latest.msi

Reviewer: we could use a datadoghq.com link instead of a s3 link: https://windows-agent.datadoghq.com/ddagent-cli-latest.msi

Author: we still use the s3 links in our in-app install instructions. Switching to the datadoghq.com link here only could make things confusing for some customers.

Reviewer: we'll keep it as is for consistency with other docs

Yes. The certificate is just a preset for the client to use and is not necessary to connect via SSL. Datadog Agent endpoints only accept SSL traffic.

### Windows Agent 5.x configured to use a proxy or the curl http client

This section applies to the Windows Agent 5.x (`<= 5.32.6`), if the Agent is configured to either:

* use a proxy with the `proxy_host` configuration option in `datadog.conf` or the `HTTPS_PROXY` environment variable, or

> * use the curl http client with the `use_curl_http_client: yes` configuration option in `datadog.conf`

Reviewer 3: ```suggestion
* use the curl HTTP client with the `use_curl_http_client: yes` configuration option in `datadog.conf`
```

Note: `datadog.conf` is located in `C:\ProgramData\Datadog\datadog.conf`.

> In this case, removing `datadog-cert.pem` will not allow the Agent to regain connectivity to Datadog. Instead, take this action:

Reviewer 3: ```suggestion
In this case, removing `datadog-cert.pem` does not allow the Agent to regain connectivity to Datadog. Instead, take this action:
```

### Additional step if the Windows Agent 5.x is configured to use a proxy or the curl http client

On Windows, with Agent 5.x, removing the certificate file alone is not enough if the Agent is configured to either:

* use a proxy with the `proxy_host` configuration option in `datadog.conf` or the `HTTPS_PROXY` environment variable, or
* use the curl http client with the `use_curl_http_client: yes` configuration option in `datadog.conf`

Note: `datadog.conf` is located in `C:\ProgramData\Datadog\datadog.conf`.

In this case, take this additional action:

* Windows Agent v5, `>= 5.12.0`: replace the `datadog-cert.pem` file with the version that is shipped in 5.32.7. Using the Powershell CLI:

```shell
Invoke-WebRequest -Uri f"https://raw.githubusercontent.com/DataDog/dd-agent/5.32.7/datadog-cert.pem" -OutFile "C:\Program Files\Datadog\Datadog Agent\agent\datadog-cert.pem"
restart-service -Force datadogagent
```

* Windows Agent v5, `<= 5.11.x`: set the following option in `datadog.conf`:
  * 64-bit Windows: `ca_certs: C:\Program Files (x86)\Datadog\Datadog Agent\files\ca-certificates.crt`

>   * 32-bit Windows: `ca_certs: C:\Program Files\Datadog\Datadog Agent\files\ca-certificates.crt`

Reviewer: related to the comment in the `Using the Windows GUI` section. Do users need to delete datadog-cert.pem too? if so, we should mention that here.

Author: (as discussed in other comment thread, the answer is that users do not need to delete the cert first)

> 

Reviewer: we should add "After datadog.conf has been updated, restart the Datadog Service from the Windows Service Manager"

Author: addressed
